### ISSUE TITLE
 Generate sourcemap for breakpoint in vscode

### ISSUE THREAD

**Issue by peter-wd-1** [netlify/netlify-lambda#262]
Hi,

First of all, I'm working on javascript project, and [here](https://github.com/netlify/cli/issues/409#issuecomment-655730329) is how I initially tried to debug lambda functions. It worked fine until I adopted ES6 syntax: `import`. This is why I started to use `netlify-lambda`. It needed to be built. 
I really think there should be a way of generating sourcemap, since [this document](https://github.com/netlify/netlify-lambda#debugging) refers to using `sourceMapPathOverrides`.  But I couldn't find a way of generating sourcemap from netlify functions.

> If using VSCode, it is likely that the sourceMapPathOverrides have to be adapted for breakpoints to work. Read here for more info on how to modify the webpack config.

> Then specify netlify-lambda serve --config ./webpack.functions.js. If using VSCode, it is likely that the sourceMapPathOverrides have to be adapted for breakpoints to work. Read here for more info on how to modify the webpack config.

It only tells me that I need it. The link on it is the same. It just a comment about how I should use webpack config file with `netlify serve` command. Also, I have no idea how turning off minify would help to debug. Does It mean that I need to add a breakpoint in a humongous build file which takes a lot of time? 

So my question is:
How to use `sourceMapPathOverrides` with netlify lambda functions? and how should I generate the sourcemap?

**Comment by michalhudecek** [comment]
Any updates on this?

**Comment by jimmend** [comment]
If your original non-compiled function code is in:

` /functions
`

then your sourceMapPathOverrides should be:

```
"sourceMapPathOverrides": {
        "webpack:///./~/*": "${workspaceFolder}/functions/*",
}

```

**Comment by michalhudecek** [comment]
This doesn't help.

**Comment by jimmend** [comment]
My comment was intended to help set the correct configuration for "sourceMapPathOverrides" once you have netlify-lambda generating a proper source map.

If you are not getting a source map, you probably need to add a --config flag to your netlify-lambda build command.

For example, instead of:

`netlify-lambda build [/location/of/function/directory]
`

use

`netlify-lambda build [/location/of/function/directory] --config [/location/of/config/file]
`

The config file should look something like this:

```
module.exports = {
  optimization: {minimize: false},
  devtool: "source-map",
  target: "node",
}

```
The key part of the config file is the 'devtool" property, which specifies which type of source map you want.  Possible values are located on this page: https://webpack.js.org/configuration/devtool/.

Now, when you run the netlify-lambda build command, it should generate a source map with the compiled function file. 

Then, you will need to make sure the sources in the source map file point to the right original source files.  If they don't, then you need to use the "sourceMapPathOverrides" setting I reference above.

Hope that helps.

**Comment by michalhudecek** [comment]
Thanks for a great explanation! I get it now. The issue is that I do not run `netlify-lambda` manually. I just run `netlify dev` and U am not sure how to tell it to run netlify-lambda with the above mentioned config.

**Comment by jimmend** [comment]
I just updated my comment.  You need a script in your package.json that references `netlify-lambda`.  `netlify dev `will find it and use the specified config and generate the source maps.

**Comment by michalhudecek** [comment]
Great! It works. The source maps are now generated. The VS Code debugger still has trouble locating them though.

I have original files in `/src/lambda` and the compiled files are in `/lambda`

My `launch.json` file looks like this:

```
{
    "version": "0.2.0",
    "configurations": [
      {
        "name": "Netlify Debugging",
        "type": "pwa-node",
        "request": "launch",
        "runtimeArgs": ["run-script","debug"],
        "runtimeExecutable": "npm",
        "skipFiles": ["<node_internals>/**"],  
        "sourceMapPathOverrides": {
          "webpack:///./~/*": "${workspaceFolder}/src/lambda/*",
        },
        "resolveSourceMapLocations": [
          "${workspaceFolder}/**",
          "!**/node_modules/**"
      ]        
      }
    ]
  }
```

And I added debug to `package.json`:
```
...
  "scripts": {
    "serve": "vue-cli-service serve",
    "build": "vue-cli-service build",
    "lint": "vue-cli-service lint",
    "start:lambda": "netlify-lambda serve src/lambda",
    "build:lambda": "netlify-lambda build src/lambda --config lambda.config.js",
    "debug": "netlify dev --inspect"
  },
...
```

Thank you

**Comment by erezrokah** [comment]
Hi @michalhudecek and @jimmend, great to see this collaboration. Can we close this issue as resolved?

**Comment by michalhudecek** [comment]
No, I am still waiting for a reply. The main issue - being able to debug lambda functions in VS Code is not resolved.

**Comment by jimmend** [comment]
Most likely, the reason that you are not hitting the breakpoints is because of the sourceMapPathOverrides setting in VSCode that I referenced in my original comment has not been set.

Let's assume you have a function file called `thisFunction.js` in the functions folder of your root directory:

`./functions/thisFunction.js`

Then, let's say, when you run `netlify dev`, the compiled functions go to `./dist/functions` (this is also where your source map will be). 

When it builds your sourcemaps, Webpack will assume that your original file is in the same directory as the sourcemap file.  If you look in your sourcemap, you will likely find this string:

`webpack:///./thisFunction.js`

What you need is for the VSCode debugger to look in `./functions` for the original file.

The way you accomplish that is be setting the sourceMapPathOverrides setting for your debugger in your launch.json:

```
{
      "name": "Attach Netlify",
      "port": 9229,
      "request": "attach",
      "skipFiles": [
        "<node_internals>/**"
      ],
      "type": "pwa-node",
      "sourceMapPathOverrides": {
        "webpack:///./*": "${workspaceFolder}/src/functions/*"
      }
}
```

Basically, what you're doing is telling the VSCode debugger to look for `thisFunction.js` in `/src/functions` instead of `/dist/functions` (or, really, for any file that it thinks is in `./dist/functions`, to look in `./src/functions`).

**Comment by jimmend** [comment]
Also, looking at your code, I think you need to change:

` "webpack:///./~/*": "${workspaceFolder}/src/lambda/*",`

to 

` "webpack:///./*": "${workspaceFolder}/src/lambda/*",`

**Comment by michalhudecek** [comment]
Thank you! That helped. It is all working now. You can close the issue :)
